src/components/user.js

```diff
@@ -15,69 +26,143 @@ export default class User extends React.Component {
+                return '';
+        }
+    }
+
+    _validateFields (user) {
+        const nameValid  = this._validateField(FIELD_NAME, user.name);
+        const phoneValid = this._validateField(FIELD_PHONE, user.phone);
+        const emailValid = this._validateField(FIELD_EMAIL, user.email);
+
+        return { nameValid, phoneValid, emailValid };
     }
 
-    handleClickAddress (e) {
+    _handleClickAddress (e) {
         this.setState({ showAddress: !this.state.showAddress });
         e.preventDefault();
     }
 
-    handleClickCompany (e) {
+    _handleClickCompany (e) {
         this.setState({ showCompany: !this.state.showCompany });
         e.preventDefault();
     }
 
-    handleClickEdit (e) {
+    _handleClickEdit (e) {
         const { currentUser, wantEdit } = this.state;
 
         if (wantEdit) {
             const newUser = new UserInfo(currentUser.info, currentUser.address, currentUser.company);
 
             UserAction.updateUser(newUser);
         }
-
         this.setState({ wantEdit: !wantEdit });
         e.preventDefault();
     }
 
-    handleChange (e, type) {
-        const { currentUser, wantEdit } = this.state;
-        const {isNewUser} = this.props;
+    _handleChange (e, obj) {
+        const { currentUser, wantEdit, formErrors } = this.state;
+        const { isNewUser }                         = this.props;
 
         if (wantEdit || isNewUser) {
-            currentUser[type][e.target.name] = e.target.value;
-            this.setState({ currentUser });
+            const name  = e.target.name;
+            const value = e.target.value;
+
+            currentUser[obj][name] = value;
+            formErrors[name]       = this._validateField(name, value);
+            this.setState({ currentUser, formErrors });
+        }
+        e.preventDefault();
+    }
+
+    _addingFailed = userId => {
+        const { isNewUser }                = this.props;
+        const { formErrors, addDuplicate } = this.state;
+
+        if (isNewUser && userId > 0) {
+            formErrors.duplicateUser.text   = ERRORS.userExist;
+            formErrors.duplicateUser.userId = `#${userId}`;
+            this.setState({ formErrors: formErrors, addDuplicate: !addDuplicate });
+        }
+    };
+
+    _handleClickSubmit (e) {
+        const { currentUser, addDuplicate, formErrors } = this.state;
+        const newUser                                   = new UserInfo(currentUser.info, currentUser.address, currentUser.company);
+        const resultValid                               = this._validateFields(currentUser.info);
+
+        if (resultValid.nameValid === '' && resultValid.phoneValid === '' && resultValid.emailValid === '') {
+            UserAction.addNewUser(newUser, addDuplicate);
+        }
+        else {
+            formErrors.name  = resultValid.nameValid;
+            formErrors.phone = resultValid.phoneValid;
+            formErrors.email = resultValid.emailValid;
+            this.setState({ formErrors });
```

Comment:
We can have an edge case situation when we set existing name -> press Submit -> seen an error
then clear a name but the error about the existing user still shown
![image](https://user-images.githubusercontent.com/12034551/96253171-ceeef580-0fbb-11eb-96f6-2ff75feec2e3.png)
But after changing email we shooul remove a dublicate error